### PR TITLE
test(shared-data): fitness spec architecture-boundaries.fitness.spec.ts

### LINKED ISSUE #246: test(shared-data): fitness spec architecture-boundaries.fitness.spec.ts
## Contexto

Defesa em profundidade behind ESLint inline. ESLint pode ser silenciado por `eslint-disable`, regras misconfiguradas ou projetos opt-out. A fitness spec re-asserta os mesmos invariantes ao nível de CI, gated por `nx run-many --target=test`.

Pattern canônico: `libs/shared-data/src/lib/__fitness__/runtime-config-providers.fitness.spec.ts` (ratificado por ADR-0021 — vitest + regex parsing de fontes).

## Cenário 1 — violação de boundary

**Dado** um `eslint.config.mjs` com `depConstraints type:*` (Story F3)
**Quando** existe um edge no graph que viola um `depConstraint`
**Então** o spec falha com a fonte e destino nomeados

## Cenário 2 — ciclo no graph

**Dado** o nx project graph
**Quando** existe um back-edge (ciclo entre projetos)
**Então** o spec falha com o caminho do ciclo

## Cenário 3 — workspace atual passa

**Dado** o workspace em `main` pós-F3
**Quando** `nx run-many --target=test --all` roda
**Então** o spec sai verde

## Implementação

**Localização:** `libs/shared-data/src/lib/__fitness__/architecture-boundaries.fitness.spec.ts`.

**Acesso ao graph:** `child_process.execSync('npx nx graph --file=/tmp/uniplus-nx-graph.json', { cwd: WORKSPACE_ROOT, stdio: 'pipe' })` em `beforeAll`, depois `JSON.parse(fs.readFileSync(...))`.

**Extração de constraints:** regex sobre `eslint.config.mjs` com mesmo comment-stripping pattern do `runtime-config-providers.fitness.spec.ts` (`/\/\*[\s\S]*?\*\//g` + `/\/\/.*$/gm`).

**Algoritmo:**

1. Para cada edge `{source, target}` em `graph.dependencies`, validar que as tags do source admitem ao menos um `depConstraint` que permita as tags do target.
2. Para detecção de ciclo, DFS por cada nó tracking visit stack — back-edge = ciclo, falha com caminho impresso.

**Performance:** budget < 5 s adicionais em `nx run-many --target=test --all`.

## Escopo

- [ ] Criar arquivo `architecture-boundaries.fitness.spec.ts` em `libs/shared-data/src/lib/__fitness__/`.
- [ ] Helpers: `extrairDepConstraints(workspaceRoot)`, `admiteDependencia(sourceTags, targetTags, constraints)`, `detectarCiclos(graph)`.
- [ ] Suite com `describe` blocks: boundary direction + acyclic graph + sanity (constraints count ≥ 11).
- [ ] Assertion guard: `expect(depConstraints.length).toBeGreaterThanOrEqual(11)` para detectar refactor que quebre o regex.
- [ ] Mutation test (off-tree): introduzir ciclo + violação deliberados, confirmar falha; nunca mergear.

## Critérios de aceite

- Spec verde em `origin/main` HEAD pós-F3.
- Mutation test registrado em comentário no PR.
- Runtime ≤ 5 s adicionais em `nx run-many --target=test --all`.

## Dependências

- Bloqueada por F3 — precisa do `depConstraints type:*` extendido para validar contra.

## Referências

- ADR-0021.
- `libs/shared-data/src/lib/__fitness__/runtime-config-providers.fitness.spec.ts` (pattern source).
- Epic #243.

### PR DESCRIPTION
**Pull request by marmota-alpina**
## Resumo

Adiciona fitness spec em \`libs/shared-data/src/lib/__fitness__/architecture-boundaries.fitness.spec.ts\` que re-asserta os mesmos invariantes de \`@nx/enforce-module-boundaries\` (introduzido em #253), mas ao nível de \`nx run-many --target=test\`. Defesa em profundidade behind o ESLint — gated por CI, imune a \`eslint-disable\`/projetos opt-out, e extra: detecta ciclos no project graph que o ESLint não cobre.

## Implementação

Pattern source: \`runtime-config-providers.fitness.spec.ts\` (ADR-0021) — vitest + regex parsing com comment-strip.

### Helpers internos

| Helper | Responsabilidade |
|---|---|
| \`carregarGraph()\` | Invoca \`nx graph --file=/tmp/...\`, parseia JSON, retorna \`{nodes, dependencies}\`. |
| \`extrairDepConstraints()\` | Lê \`eslint.config.mjs\`, strip comentários, regex \`sourceTag: '...', onlyDependOnLibsWithTags: [...]\`. |
| \`admiteDependencia(sourceTags, targetTags, constraints)\` | Para cada constraint aplicável, target precisa ter ao menos uma tag em \`onlyDependOnLibsWithTags\`. |
| \`detectarCiclos(graph)\` | DFS com \`visited + active\` sets; back-edge = ciclo; retorna lista de caminhos. |

### Specs

| describe / it | Garante |
|---|---|
| \`extrai ao menos 11 entradas\` | Sanity guard: refactor que reduza ou quebre o regex invalida o gate. |
| \`todo edge do project graph respeita as constraints\` | Itera \`graph.dependencies\` e aplica \`admiteDependencia\` em cada source → target. |
| \`graph é acíclico\` | DFS sobre todos os nodes; falha imprime o caminho do ciclo. |

## Validação

| Gate | Resultado |
|---|---|
| \`nx run shared-data:test\` | 8/8 verde, 84 testes; spec rodou em **606 ms** (budget < 5 s da Story) |
| \`nx run-many --target=test --all\` | 8/8 verde, 7 s wall-clock |
| \`nx run-many --target=lint --all\` | 12/12 verde |
| \`nx run-many --target=build --all\` | 8/8 verde |

### Mutation test off-tree

Inject de \`import { EditalDto } from '@uniplus/shared-data'\` em arquivo temporário \`libs/shared-core/src/lib/_mutation-violation.ts\`. Disparou **as 2 specs simultaneamente** (boundary + ciclo):

\`\`\`
✗ todo edge do project graph respeita as constraints

  Violações de boundary detectadas:
    shared-core [npm:public, scope:shared, type:core] -> shared-data [npm:public, scope:shared, type:data]
    viola constraint sourceTag=type:core (precisa ao menos uma de [])

✗ graph é acíclico (sem back-edges entre projects)

  (back-edge detectado: shared-data já dependia de shared-core; mutation criou
   shared-core -> shared-data, fechando o ciclo)
\`\`\`

Branch nunca commitada; arquivo removido após validação. Cobre o edge case \`type:core\` leaf (\`onlyDependOnLibsWithTags: []\`) levantado como S2 na review de #253.

## Critérios de aceite

- [x] Spec verde em \`origin/main\` HEAD pós-#245.
- [x] Mutation test registrado.
- [x] Runtime ≤ 5 s adicionais em \`nx run-many --target=test --all\` (606 ms observado).

## Riscos / rollback

Risco baixo — spec puramente declarativo, não toca runtime. Rollback: \`git revert\`.

Closes #246

Refs Epic #243

### CODE REVIEW

File: libs/shared-data/src/lib/__fitness__/architecture-boundaries.fitness.spec.ts [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+  const aplicaveis = constraints.filter((c) => sourceTags.includes(c.sourceTag));
+  for (const c of aplicaveis) {
+    const satisfaz = targetTags.some((t) =>
+      c.onlyDependOnLibsWithTags.includes(t),
+    );
+    if (!satisfaz) {
+      return { ok: false, constraint: c };
+    }
+  }
+  return { ok: true, constraint: null };
```
[the remark this answers is not in the document]
Comment (marmota-alpina) [reply libs/shared-data/src/lib/__fitness__/architecture-boundaries.fitness.spec.ts:114]: Bom achado, P1 absorvido em `aed473b`.

`admiteDependencia()` agora trata `aplicaveis.length === 0` como `{ ok: false, motivo: 'sem_constraint_aplicavel' }`. Fitness fica mais estrito que `@nx/enforce-module-boundaries` aqui — qualquer lib mistagged ou novo project sem tagging adequado falha o gate, com mensagem de erro guiando o ajuste (adicionar entrada em depConstraints ou corrigir as tags do project). Apps `type:app` e e2e `type:e2e` continuam passando porque carregam também `scope:*` (que tem entrada na matriz).

File: libs/shared-data/src/lib/__fitness__/architecture-boundaries.fitness.spec.ts [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+      ).toBeGreaterThanOrEqual(11);
+    });
+
+    it('todo edge do project graph respeita as constraints', () => {
+      const violacoes: string[] = [];
+      for (const [source, edges] of Object.entries(graph.dependencies)) {
+        const sourceTags = graph.nodes[source]?.data.tags ?? [];
+        for (const edge of edges) {
+          if (edge.target.startsWith('npm:')) continue;
+          const targetNode = graph.nodes[edge.target];
```
[the remark this answers is not in the document]
Comment (marmota-alpina) [reply libs/shared-data/src/lib/__fitness__/architecture-boundaries.fitness.spec.ts:183]: P2 absorvido em `aed473b`.

Adicionei `if (edge.type === 'implicit') continue;` antes da validação. Edges `implicit` (declaradas via `implicitDependencies` em `apps/*-e2e/project.json`) representam relação operacional Nx, não import statement TypeScript — ESLint não enforça constraints sobre elas, então o fitness também não deve enforçar para preservar a paridade "mesmos invariantes".

**Review by jf2s — APPROVED** [review]
Aprovado.

Inspecionado:

- Codex P1 (admiteDependencia retornava ok:true para source sem matching constraint) absorvido em `aed473b` — agora retorna `{ ok: false, motivo: 'sem_constraint_aplicavel' }` com mensagem guiando o ajuste. Apps `type:app` e e2e `type:e2e` continuam passando porque carregam também `scope:*` (que tem entrada na matriz).
- Codex P2 (graph inclui implicit edges via `implicitDependencies` em apps/*-e2e/project.json) absorvido em `aed473b` — `if (edge.type === 'implicit') continue;` antes da validação preserva paridade com ESLint que só valida imports TS.
- Threads do Codex (P1 + P2) resolvidas via GraphQL.
- CI 5/5 SUCCESS após fix push.
- Codex AI re-react 👍 sem novos findings.
- Suite completa: 8/8 verde, 84 testes; spec architecture-boundaries em 551 ms (budget < 5 s).
- Mutation test off-tree do commit inicial confirmou que o spec detecta boundary violation + ciclo simultaneamente.
